torsiondrive/qm_engine.py

```diff
@@ -327,6 +327,151 @@ def load_native_output(self, filename='output.dat'):
+                    elems.append(ls[0])
+                    coords.append(ls[1:])
+                    if not found_geo:
+                        found_geo = True
+                        gauss_temp.append("$!geometry@here")
+
+                elif reading_molecule:
+                    if line.strip() == '':
+                        reading_molecule = False
+                        gauss_temp.append(line)
+                        gauss_temp.append("$!optblock@here")
+
+                else:
+                    gauss_temp.append(line)
+
+                if 'opt' in line.lower() and 'modredundant' in line.lower():
+                    self.temp_type = 'optimize'
+        assert found_geo, "XYZ geometry not found in molecule block of %s" % input_file
+        if self.native_opt:
+            assert self.temp_type == 'optimize', "input_file should be a opt job to use native opt"
+        self.gauss_temp = gauss_temp
+        self.M = Molecule()
+        self.M.elem = elems
+        self.M.xyzs = [np.array(coords, dtype=float)]
+        self.M.build_topology()
+
+    def write_input(self, filename='gaussian.com'):
+        """ Write Gaussian input using Molecule Class """
+        assert hasattr(self, 'gauss_temp'), "self.gauss_temp not set, call load_input() first"
+        with open(filename, 'w') as outfile:
+            for line in self.gauss_temp:
+                if line == '$!geometry@here':
+                    for e, c in zip(self.M.elem, self.M.xyzs[0]):
+                        outfile.write("%-7s %13.7f %13.7f %13.7f\n" % (e, c[0], c[1], c[2]))
+
+                elif line == "$!optblock@here":
+                    if hasattr(self, 'optblockStr'):
+                        # self.optblockStr will be set by self.optimize_native()
+                        outfile.write(self.optblockStr)
+
+                else:
+                    outfile.write(line)
+
+    def optimize_native(self):
+        """
+        Run the constrained optimization, following Gaussian09 manual.
+        1. write a optimization job input file.
+        2. run the job
+        """
+        assert self.temp_type == 'optimize', "To use native optimization, the input file be an opt job"
+        assert hasattr(self, 'gaussian_version'), 'The version of gaussian could not be determined!'
+        if self.extra_constraints is not None:
+            raise RuntimeError('extra constraints not supported in Gaussian09 native optimizations')
+        self.optblockStr =''
+        for d1, d2, d3, d4, v in self.dihedral_idx_values:
+            self.optblockStr += f'{d1 + 1} {d2 + 1} {d3 + 1} {d4 + 1} ={v:.3f} B\n'  # Build the angle
+            self.optblockStr += f'{d1 + 1} {d2 + 1} {d3 + 1} {d4 + 1} F\n'           # Freeze the angle
+        # write input file
+        self.write_input('gaussian.com')
+        # run the job
+        self.run(f'{self.gaussian_version} < gaussian.com > gaussian.log && formchk lig.chk lig.fchk', input_files=['gaussian.com'],
```

Comment:
Renaming `self.gaussian_version` to `self.gaussian_exe` will make it a little easier to read I think.